source/common/router/config_impl.cc

```diff
@@ -85,11 +85,45 @@ RouteEntryImplBase::RouteEntryImplBase(const VirtualHostImpl& vhost, const Json:
       rate_limit_policy_(route), shadow_policy_(route),
       priority_(ConfigUtility::parsePriority(route)) {
 
+  bool have_weighted_clusters = route.hasObject("weighted_clusters");
+  bool have_cluster = !cluster_name_.empty() || have_weighted_clusters;
   // Check to make sure that we are either a redirect route or we have a cluster.
-  if (!(isRedirect() ^ !cluster_name_.empty())) {
+  if (!(isRedirect() ^ have_cluster)) {
     throw EnvoyException("routes must be either redirects or cluster targets");
   }
 
+  if (have_cluster) {
+    if (!cluster_name_.empty() && have_weighted_clusters) {
+      throw EnvoyException("routes must have either single or weighted_clusters as target");
+    }
+  }
+
+  // If this is a weighted_cluster, we create N internal route entries
+  // (called WeightedClusterEntry), such that each object is a simple
+  // single cluster, pointing back to the parent.
+  if (have_weighted_clusters) {
+    uint64_t total_weight = 0UL, max_weight = 100UL; // TODO: move this constant to a header file
+
+    const std::vector<Json::ObjectPtr> weighted_clusters_json =
+        route.getObjectArray("weighted_clusters");
+
+    for (const Json::ObjectPtr& cluster : weighted_clusters_json) {
+      // Internal route entry objects, one per weighted cluster entry in the config
+      weighted_clusters_.emplace_back(std::make_shared<WeightedClusterEntry>(
```

Comment:
This should be unique_ptr. I just looked and I'm not sure why RouteEntryImplBasePtr is a shared_ptr currently. Switch that to unique_ptr also.